README
NAME
    POE::Filter::IRCv3 - POE::Filter::IRCD with IRCv3.2 message tags

SYNOPSIS
      my $filter = IRC::Server::Pluggable::IRC::Filter->new(colonify => 1);

      ## Raw lines parsed to hashes:
      my $array_of_refs  = $filter->get( [ $line1, $line ... ] );

      ## Hashes deparsed to raw lines:
      my $array_of_lines = $filter->put( [ \%hash1, \%hash2 ... ] );

      ## Stacked with a line filter, suitable for Wheel usage, etc:
      my $ircd = IRC::Server::Pluggable::IRC::Filter->new(colonify => 1);
      my $line = POE::Filter::Line->new(
        InputRegexp   => '\015?\012',
        OutputLiteral => "\015\012",
      );
      my $filter = POE::Filter::Stackable->new(
        Filters => [ $line, $ircd ],
      );

DESCRIPTION
    A POE::Filter for IRC traffic derived from POE::Filter::IRCD.

    Adds support for IRCv3.2 message tags.

    Like any proper POE::Filter, there are no POE-specific bits involved
    here; the filter can be used stand-alone to parse IRC traffic.

  get_one_start, get_one, get_pending
    Implement the interface described in POE::Filter.

    See "get".

  get
      my $events = $filter->get( [ $line, $another, ... ] );
      for my $event (@$events) {
        my $cmd = $event->{command};
        ## See below for other keys available
      }

    Takes an ARRAY of raw lines and returns an ARRAY of hash references with
    the following keys:

   command
    The (uppercased) command or numeric.

   params
    An ARRAY containing the event parameters.

   prefix
    The sender prefix, if any.

   tags
    A HASH of key => value pairs matching IRCv3.2 "message tags" -- see
    <http://ircv3.atheme.org>.

    Note that a tag can be present, but have an undefined value.

  put
      my $lines = $filter->put( [ $hash, $another_hash, ... ] );
      for my $line (@$lines) {
        ## Direct to socket, etc
      }

    Takes an ARRAY of hash references matching those described in "get"
    (documented above) and returns an ARRAY of raw IRC-formatted lines.

   colonify
    In addition to the keys described in "get", the colonify option can be
    specified for specific events. This controls whether or not the last
    parameter will be colon-prefixed even if it is a single word. (Yes, IRC
    is woefully inconsistent ...)

    Defaults to boolean false (off).

  clone
    Copy the filter object (with a cleared buffer).

  debug
    Turn on/off debug output.

AUTHOR
    Derived from POE::Filter::IRCD, which is copyright Chris Williams and
    Jonathan Steinert.

    Adapted with IRCv3 extensions and other tweaks by Jon Portnoy
    <avenj@cobaltirc.org>

    This module may be used, modified, and distributed under the same terms
    as Perl itself. Please see the license that came with your Perl
    distribution for details.

SEE ALSO
    POE::Filter

    POE::Filter::IRCD

    POE::Filter::Line

    POE::Filter::Stackable

    IRC::Toolkit



Changes
Revision history for Perl module POE::Filter::IRCv3

1.002001 2015-04-23

  - Remove '\0' escaping in tags, per ircv3-specifications GH #128

  - Drop fatal warnings

  - Switch to 'cpanfile' packaging

1.001001 2014-08-16

  - Single trailing param starting with ':' should prepend a ':'
    Related bug in JS 'irc-message' (derived from this dist):
    https://github.com/expr/irc-message/pull/12

  - Minor optimizations

  - Happy CPAN Day!

0.047002 2014-06-20

  - Message tag escape/unescape optimizations

0.047001 2014-06-20

  - Escape message tag values per ircv3-specifications issue #71

  - Test coverage improvements

0.046000 2013-12-06

  - Allow public access to the underlying parse function;
    useful for faster standalone parsing or custom failure handling

0.045004 2013-12-02

  - Drop 'strictures' & 'parent' dependencies

  - POD / test improvements

0.045003 2013-10-26

  - POE dependency is now optional

0.045002 2013-08-10

  - No code changes; fix 'Changes' to match CPAN::Changes::Spec

0.045001 2013-06-15

  - Minor optimizations.

0.045000 2013-06-12

  - Swap params parser implementations; slightly faster on strings containing
    many middle parameters.

  - Turn on ->debug(1) at construction time if env POE_FILTER_IRC_DEBUG is 
    true.

0.044000 2013-06-09

  - Fix a bug in detecting invalid zero-length prefix; tests for same.

0.043000 2013-06-09

  - Fix a bug in detecting extraneous spaces between middle and trailing 
    params.

  - New test suite and improved helpers for same.

0.042003 2013-06-08

  - POD fixes; leftovers from ancient import (how did that get there?)

0.042002 2013-06-08

  - Fix trailing spaces in commands; previously returned an empty
    string parameter in ->{params}

0.042001 2013-06-07

  - Fix per-event 'colonify =>' configurable; tests for same

0.042000 2013-06-07

  - Improved non-regex parser and tests for same.

0.041001 2013-06-05

  - Revert to previous parser until we have better tests

0.041000 2013-06-05

  - Switch to a non-regex get_one() implementation.
    Benching shows this is faster in many cases.

0.040001 2013-02-04

  - Empty-string prefix should DWIM

  - Switch to long-format semantic version

0.04 2013-02-04

  - Missing dep on parent . . . again.

0.03 2013-02-1

  - Fix a broken test (5.17 exposed hash order mistake)

0.02 2013-01-31

  - Missing dependency on 'parent'

0.01 2013-01-27

  - Initial release


t/inc/TestFilterHelpers.pm
package TestFilterHelpers;
use strict; use warnings FATAL => 'all';

require Carp;
require Scalar::Util;

=pod

=head1 NAME

TestFilterHelpers - POE::Filter::IRC(d,v3) test helpers

=head1 SYNOPSIS

  use Test::More;
  use lib 't/inc';
  use TestFilterHelpers;

  my $line = ':test foo';
  get_ok $filter, $line =>
    +{
        raw_line => $line,
        command  => 'FOO',
        prefix   => 'test',
    },
    'my get test ok' ;

  put_ok $filter, $line =>
    +{ command => 'foo', prefix => 'test' },
    'my put test ok' ;

  get_command_ok $filter, $line => $cmd, $name;

  get_prefix_ok $filter, $line => $prefix, $name;

  get_params_ok $filter, $line => [@params], $name;

  get_rawline_ok $filter, $line, $name;

  get_tags_ok $filter, $line => +{%tags}, $name;

  done_testing;

=head1 DESCRIPTION

A simple set of helpers for testing L<POE::Filter::IRCv3> (and
L<POE::Filter::IRCD>).

=head1 AUTHOR

Jon Portnoy <avenj@cobaltirc.org>

=cut


use Test::Deep::NoTest qw/
  cmp_deeply

  cmp_details
  deep_diag
/;

use base 'Exporter';
our @EXPORT = qw/
  get_ok
  put_ok

  get_command_ok
  get_prefix_ok
  get_params_ok
  get_rawline_ok
  get_tags_ok
/;

my $Test = Test::Builder->new;
sub import {
  my $self = shift;
  if (@_) {
    my $pkg = caller;
    $Test->exported_to( $pkg );
    $Test->plan( @_ );
  }
  $self->export_to_level( 1, $self, $_ ) for @EXPORT;
}


sub _looks_ok {
  my ($got, $expected, $name) = @_;

  local $Test::Builder::Level = $Test::Builder::Level + 1;

  my ($ok, $stack) = cmp_details($got, $expected);

  unless ( $Test->ok($ok, $name) && return 1 ) {
    if (ref $got && ref $expected) {
      $Test->diag( "Structures:\n",
        "Expected ->\n",
        $Test->explain($expected),
        "Got ->\n",
        $Test->explain($got),
      )
    }

    $Test->diag( deep_diag($stack) )
  }

  return
}


sub put_ok {
  my ($filter, $line, $ref, $name) = @_;

  unless (Scalar::Util::blessed $filter) {
    Carp::croak "put_ok expected blessed filter obj"
  }

  unless (defined $line && !ref($line) && ref $ref eq 'HASH') {
    Carp::croak "put_ok expected a line to compare and a HASH to process"
  }

  my $arr = $filter->put([ $ref ]);
  Carp::croak "filter did not return ARRAY for $ref"
    unless ref $arr eq 'ARRAY';
  Carp::croak "filter did not return line for $ref"
    unless defined $arr->[0];

  $name = 'line looks ok' unless defined $name;

  _looks_ok( $arr, [ $line ], $name ) ? $arr->[0] : ()
}


sub get_ok {
  my ($filter, $line, $ref, $name) = @_;

  unless (Scalar::Util::blessed $filter) {
    Carp::croak "get_ok expected blessed filter obj"
  }

  unless (defined $line && ref $ref eq 'HASH') {
    Carp::croak "get_ok expected a line to process and HASH to compare"      
  }

  $ref->{raw_line} = $line unless exists $ref->{raw_line};

  my $arr = $filter->get([ $line ]);

  Carp::croak "filter did not return ARRAY for $line"
    unless ref $arr eq 'ARRAY';

  $name = 'struct looks ok' unless defined $name;
  _looks_ok( $arr, [$ref], $name ) ? $arr->[0] : ()
}


sub get_command_ok {
  my ($filter, $line, $cmd, $name) = @_;

  unless (Scalar::Util::blessed $filter) {
    Carp::croak "get_command_ok expected blessed filter obj"
  }

  unless (defined $line && defined $cmd) {
    Carp::croak "get_command_ok expected a line to process and command to compare"
  }

  my $arr = $filter->get([ $line ]);

  Carp::croak "filter did not return ARRAY for $line"
    unless ref $arr eq 'ARRAY';
  Carp::croak "filter did not return event for $line"
    unless ref $arr->[0] eq 'HASH';

  $name = 'command looks ok' unless defined $name;
  _looks_ok( $arr->[0]->{command}, $cmd, $name )
}

sub get_prefix_ok {
  my ($filter, $line, $pfx, $name) = @_;

  unless (Scalar::Util::blessed $filter) {
    Carp::croak "get_prefix_ok expected blessed filter obj"
  }

  # undef prefix is a valid comparison:
  unless (defined $line) {
    Carp::croak "get_prefix_ok expected a line to process and prefix to compare"
  }

  my $arr = $filter->get([ $line ]);
  Carp::croak "filter did not return ARRAY for $line"
    unless ref $arr eq 'ARRAY';
  Carp::croak "filter did not return event for $line"
    unless ref $arr->[0] eq 'HASH';

  $name = 'prefix looks ok' unless defined $name;
  _looks_ok( $arr->[0]->{prefix}, $pfx, $name )
}

sub get_params_ok {
  my ($filter, $line, $pref, $name) = @_;

  unless (Scalar::Util::blessed $filter) {
    Carp::croak "get_params_ok expected blessed filter obj"
  }

  # pref => undef is legit
  unless (defined $line) {
    Carp::croak "get_params_ok expected a line to process and params to compare"
  }

  my $arr = $filter->get([ $line ]);
  Carp::croak "filter did not return ARRAY for $line"
    unless ref $arr eq 'ARRAY';
  Carp::croak "filter did not return event for $line"
    unless ref $arr->[0] eq 'HASH';

  $name = 'params look ok' unless defined $name;
  _looks_ok( $arr->[0]->{params}, $pref, $name )
}

sub get_rawline_ok {
  my ($filter, $line, $name) = @_;
  
  unless (Scalar::Util::blessed $filter) {
    Carp::croak "get_rawline_ok expected blessed filter obj"
  }

  unless (defined $line) {
    Carp::croak "get_rawline_ok expected a line to process"
  }

  my $arr = $filter->get([ $line ]);
  Carp::croak "filter did not return ARRAY for $line"
    unless ref $arr eq 'ARRAY';
  Carp::croak "filter did not return event for $line"
    unless ref $arr->[0] eq 'HASH';

  $name = 'raw_line looks ok' unless defined $name;
  _looks_ok( $arr->[0]->{raw_line}, $line, $name )
}

sub get_tags_ok {
  my ($filter, $line, $tags, $name) = @_;

  unless (Scalar::Util::blessed $filter) {
    Carp::croak "get_tags_ok expected blessed filter obj"
  }

  unless (defined $line) {
    Carp::croak "get_tags_ok expected a line to process and a tags HASH to compare"
  }

  my $arr = $filter->get([ $line ]);
  Carp::croak "filter did not return ARRAY for $line"
    unless ref $arr eq 'ARRAY';
  Carp::croak "filter did not return event for $line"
    unless ref $arr->[0] eq 'HASH';

  $name = 'tags look ok' unless defined $name;
  _looks_ok( $arr->[0]->{tags}, $tags, $name )
}


1;


bench/vs_poe_filter_ircd.pl
use strictures 1;
use Benchmark ':all';


require POE::Filter::IRCD;
require POE::Filter::IRCv3;

my ($old, $new);
print " -> object construction <- \n";
cmpthese( 500_000, +{
  ircv3  => sub { $new = POE::Filter::IRCv3->new },
  ircd   => sub { $old = POE::Filter::IRCD->new  }
});

{
  my $basic = ':test!me@test.ing PRIVMSG #Test :This is a test'
              .' but not of the emergency broadcast system'
              .' foo bar baz quux snarf';

  print " -> common string get() <- \n";

  my ($ev_new, $ev_old);
  cmpthese( 500_000, +{
    ircv3  => sub { $ev_new = $new->get([ $basic ]) },
    ircd   => sub { $ev_old = $old->get([ $basic ]) }
  });


  print " -> common string put() <- \n";

  cmpthese( 500_000, +{
    ircv3 => sub { my $lines = $new->put([ @$ev_new ]) },
    ircd  => sub { my $lines = $old->put([ @$ev_old ]) }
  });
}

{
  my $multiparam = ':test!me@test.ing foo bar baz quux frobulate snack';

  print " -> multi-param string get() <- \n";

  my ($ev_new, $ev_old);
  cmpthese( 500_000, +{
    ircv3  => sub { $ev_new = $new->get([ $multiparam ]) },
    ircd   => sub { $ev_old = $old->get([ $multiparam ]) }
  });


  print " -> multi-param string put() <- \n";

  cmpthese( 500_000, +{
    ircv3 => sub { my $lines = $new->put([ @$ev_new ]) },
    ircd  => sub { my $lines = $old->put([ @$ev_old ]) }
  });
}



lib/POE/Filter/IRCv3/NoRE.pm
package POE::Filter::IRCv3::NoRE;

use strictures 1;
use Carp;

use parent 'POE::Filter';

=pod

=for Pod::Coverage COLONIFY DEBUG BUFFER

=cut

sub COLONIFY () { 0 }
sub DEBUG    () { 1 }
sub BUFFER   () { 2 }


sub new {
  my ($class, %params) = @_;
  $params{uc $_} = $params{$_} for keys %params;

  my $self = [
    $params{'COLONIFY'} || 0,
    $params{'DEBUG'}    || 0,
    []  ## BUFFER
  ];

  bless $self, $class;
}

sub clone {
  my ($self) = @_;
  my $nself = [@$self];
  $nself->[BUFFER] = [];
  bless $nself, ref $self
}

sub debug {
  my ($self, $value) = @_;
  return $self->[DEBUG] = $value if defined $value;
  $self->[DEBUG]
}

sub colonify {
  my ($self, $value) = @_;
  return $self->[COLONIFY] = $value if defined $value;
  $self->[COLONIFY]
}

sub get_one_start {
  my ($self, $raw_lines) = @_;
  push @{ $self->[BUFFER] }, $_ for @$raw_lines;
}

sub get_pending {
  my ($self) = @_;
  @{ $self->[BUFFER] } ? [ @{ $self->[BUFFER] } ] : ()
}

sub _parseline {
  my ($raw_line) = @_;
  my @input = split ' ', ( $raw_line || return );
  my %event = ( raw_line => $raw_line );

  if ( index($input[0], '@') == 0 ) {
    for my $tag_pair (split /;/, substr $input[0], 1) {
      my ($thistag, $thisval) = split /=/, $tag_pair;
      $event{tags}->{$thistag} = $thisval
    }
    shift @input
  }

  if ( index($input[0], ':') == 0 ) {
    $event{prefix} = substr $input[0], 1;
    shift @input
  }

  $event{command} = uc( shift(@input) || return );

  PARAM: while (defined (my $param = shift @input)) {
    if ( index($param, ':') == 0 ) {
      push @{ $event{params} }, join ' ', substr($param, 1), @input;
      last PARAM
    }
    push @{ $event{params} }, $param;
  }

  \%event
}

sub get_one {
  my ($self) = @_;
  my @events;

  if (my $raw_line = shift @{ $self->[BUFFER] }) {
    warn "-> $raw_line \n" if $self->[DEBUG];

    if (my $event = _parseline($raw_line)) {
      push @events, $event;
    } else {
      carp "Received malformed IRC input: $raw_line";
    }
  }

  \@events
}

sub put {
  my ($self, $events) = @_;
  my $raw_lines = [];

  for my $event (@$events) {

    if ( ref $event eq 'HASH' ) {
      my $raw_line;

      if ( ref $event->{tags} eq 'HASH' && keys %{ $event->{tags} } ) {
          $raw_line .= '@';
          my @tags = %{ $event->{tags} };
          while (my ($thistag, $thisval) = splice @tags, 0, 2) {
            $raw_line .= $thistag . ( defined $thisval ? '='.$thisval : '' );
            $raw_line .= ';' if @tags;
          }
          $raw_line .= ' ';
      }

      $raw_line .= ':' . $event->{prefix} . ' ' if $event->{prefix};
      $raw_line .= $event->{command};

      if ( ref $event->{params} eq 'ARRAY'
        && (my @params = @{ $event->{params} }) ) {
          $raw_line .= ' ';
          my $param = shift @params;
          while (@params) {
            $raw_line .= $param . ' ';
            $param = shift @params;
          }
          $raw_line .= ':'
            if $self->[COLONIFY]
            or $event->{colonify}
            or $param =~ m/\x20/;
          $raw_line .= $param;
      }

      push @$raw_lines, $raw_line;
      warn "<- $raw_line \n" if $self->[DEBUG];
    } else {
      carp "($self) non-HASH passed to put(): '$event'";
      push @$raw_lines, $event if ref $event eq 'SCALAR';
    }

  }

  $raw_lines
}

1;


=pod

=head1 NAME

POE::Filter::IRCv3 - POE IRC filter IRCv3.2 message tags

=head1 SYNOPSIS

  my $filter = IRC::Server::Pluggable::IRC::Filter->new(colonify => 1);

  ## Raw lines parsed to hashes:
  my $array_of_refs  = $filter->get( [ $line1, $line ... ] );

  ## Hashes deparsed to raw lines:
  my $array_of_lines = $filter->put( [ \%hash1, \%hash2 ... ] );

  ## Stacked with a line filter, suitable for Wheel usage, etc:
  my $ircd = IRC::Server::Pluggable::IRC::Filter->new(colonify => 1);
  my $line = POE::Filter::Line->new(
    InputRegexp   => '\015?\012',
    OutputLiteral => "\015\012",
  );
  my $filter = POE::Filter::Stackable->new(
    Filters => [ $line, $ircd ],
  );

=head1 DESCRIPTION

A L<POE::Filter> for IRC traffic.

Adds support for IRCv3.2 message tags.

Like any proper L<POE::Filter>, there are no POE-specific bits involved 
here; the filter can be used stand-alone to parse IRC traffic (see
L<IRC::Toolkit::Parser>).

=head2 new

Construct a new Filter; if the B<colonify> option is true, 
the last parameter will always have a colon prepended.
(This setting can also be retrieved or changed on-the-fly by calling 
B<colonify> as a method, or changed for specific events by passing a 
B<colonify> option via events passed to L</put>.)

=head2 get_one_start, get_one, get_pending

Implement the interface described in L<POE::Filter>.

See L</get>.

=head2 get

  my $events = $filter->get( [ $line, $another, ... ] );
  for my $event (@$events) {
    my $cmd = $event->{command};
    ## See below for other keys available
  }

Takes an ARRAY of raw lines and returns an ARRAY of HASH-type references with 
the following keys:

=head3 command

The (uppercased) command or numeric.

=head3 params

An ARRAY containing the event parameters.

=head3 prefix

The sender prefix, if any.

=head3 tags

A HASH of key => value pairs matching IRCv3.2 "message tags" -- see 
L<http://ircv3.atheme.org>.

Note that a tag can be present, but have an undefined value.

=head2 put

  my $lines = $filter->put( [ $hash, $another_hash, ... ] );
  for my $line (@$lines) {
    ## Direct to socket, etc
  }

Takes an ARRAY of HASH-type references matching those described in L</get> 
(documented above) and returns an ARRAY of raw IRC-formatted lines.

=head3 colonify

In addition to the keys described in L</get>, the B<colonify> option can be 
specified for specific events. This controls whether or not the last 
parameter will be colon-prefixed even if it is a single word. (Yes, IRC is 
woefully inconsistent ...)

Specify as part of the event hash:

  $filter->put([ { %event, colonify => 1 } ]);

=head2 clone

Copy the filter object (with a cleared buffer).

=head2 debug

Turn on/off debug output.

=head1 AUTHOR

Jon Portnoy <avenj@cobaltirc.org>

Originally derived from L<POE::Filter::IRCD>, which is copyright Chris
Williams and Jonathan Steinert; heavily modified since that time, such that
little original code remains, but I stand on the shoulders of giants ;-)

Licensed under the same terms as Perl.

=head1 SEE ALSO

L<POE::Filter>

L<POE::Filter::IRCD>

L<POE::Filter::Line>

L<POE::Filter::Stackable>

L<IRC::Toolkit>

=cut


lib/POE/Filter/IRCv3.pm
package POE::Filter::IRCv3;

use strict; use warnings;

use Carp;

# Not mandatory, but handy for POE apps, in which case POE is presumably
# installed already; we can do all POE::Filter-y things, this only matters for
# ->isa('POE::Filter') ->
BEGIN {
  if (eval { require POE::Filter; 1 }) {
    our @ISA = 'POE::Filter';
  }
}

=pod

=for Pod::Coverage COLONIFY DEBUG BUFFER SPCHR

=cut

sub COLONIFY () { 0 }
sub DEBUG    () { 1 }
sub BUFFER   () { 2 }
sub SPCHR    () { "\x20" }

our %CharToEscapedTag = (
  ';'  => '\:',
  ' '  => '\s',
  "\\" => '\\',
  "\r" => '\r',
  "\n" => '\n',
  "\a" => '\a',
);

our %EscapedTagToChar = reverse %CharToEscapedTag;

sub new {
  my ($class, %params) = @_;
  $params{uc $_} = $params{$_} for keys %params;
  bless [
    ($params{'COLONIFY'} || 0),
    ($params{'DEBUG'}    || $ENV{POE_FILTER_IRC_DEBUG} || 0),
    []      # BUFFER
  ], $class
}

sub clone {
  my ($self) = @_;
  my $nself = [@$self];
  $nself->[BUFFER] = [];
  bless $nself, ref $self
}

sub debug    { defined $_[1] ? $_[0]->[DEBUG] = $_[1] : $_[0]->[DEBUG] }
sub colonify { defined $_[1] ? $_[0]->[COLONIFY] = $_[1] : $_[0]->[COLONIFY] }

sub get_one_start { push @{ $_[0]->[BUFFER] }, $_ for @{ $_[1] }; }
sub get_pending   { @{ $_[0]->[BUFFER] } ? [ @{ $_[0]->[BUFFER] } ] : () }

sub get {
  my @events;
  for my $raw_line (@{ $_[1] }) {
    warn " >> '$raw_line'\n" if $_[0]->[DEBUG];
    if ( my $event = parse_one_line($raw_line) ) {
      push @events, $event;
    } else {
      carp "Received malformed IRC input: $raw_line";
    }
  }
  \@events
}

sub get_one {
  my ($self) = @_;
  my @events;
  if ( my $raw_line = shift @{ $self->[BUFFER] } ) {
    warn " >> '$raw_line'\n" if $self->[DEBUG];
    if ( my $event = parse_one_line($raw_line) ) {
      push @events, $event;
    } else {
      warn "Received malformed IRC input: $raw_line\n";
    }
  }
  \@events
}


use bytes;
no warnings 'substr';

sub put {
  my ($self, $events) = @_;
  my $raw_lines = [];

  for my $event (@$events) {

    if ( ref $event eq 'HASH' ) {
      my $raw_line;

      ## FIXME this gets glacially slow ->
      #  y// over string for escapes first and then loop?
      #  
      if ( exists $event->{tags} && (my @tags = %{ $event->{tags} }) ) {
          $raw_line .= '@';
          while (my ($thistag, $thisval) = splice @tags, 0, 2) {
            $raw_line .= $thistag;
            if (defined $thisval) {
              $raw_line .= '=';
              my $tag_pos = 0;
              my $len = length $thisval;
              while ($tag_pos < $len) {
                my $ch = substr $thisval, $tag_pos++, 1;
                $raw_line .= exists $CharToEscapedTag{$ch} ?
                  $CharToEscapedTag{$ch} : $ch
              }
            }
            $raw_line .= ';' if @tags;
          }
          $raw_line .= ' ';
      }

      $raw_line .= ':' . $event->{prefix} . ' ' if $event->{prefix};
      $raw_line .= $event->{command};

      if ( $event->{params} && (my @params = @{ $event->{params} }) ) {
          $raw_line .= ' ';
          my $param = shift @params;
          while (@params) {
            $raw_line .= $param . ' ';
            $param = shift @params;
          }
          $raw_line .= ':'
            if (index($param, SPCHR) != -1)
            or (index($param, ':') == 0)
            or (
              defined $event->{colonify} ?
              $event->{colonify} : $self->[COLONIFY]
            );
          $raw_line .= $param;
      }

      push @$raw_lines, $raw_line;
      warn " << '$raw_line'\n" if $self->[DEBUG];
    } else {
      carp "($self) non-HASH passed to put(): '$event'";
      push @$raw_lines, $event if ref $event eq 'SCALAR';
    }

  }

  $raw_lines
}


sub parse_one_line {
  my $raw_line = $_[0];
  my %event = ( raw_line => $raw_line );
  my $pos = 0;

  ## We cheat a little; the spec is fuzzy when it comes to CR, LF, and NUL
  ## bytes. Theoretically they're not allowed inside messages, but
  ## that's really an implementation detail (and the spec agrees).
  ## We just stick to SPCHR (\x20) here.

  if ( substr($raw_line, 0, 1) eq '@' ) {
    return unless (my $nextsp = index($raw_line, SPCHR)) > 0;
    # Tag parser cheats and uses split, at the moment:
    for ( split /;/, substr $raw_line, 1, ($nextsp - 1) ) {
          my ($thistag, $thisval) = split /=/;
          my $realval;
          if (defined $thisval) {
            my $tag_pos = 0;
            my $len = length $thisval;
            while ($tag_pos < $len) {
              my $ch = substr $thisval, $tag_pos++, 1;
              if ($ch eq "\\") {
                my $pair = $ch . (substr $thisval, $tag_pos++, 1 || '');
                $realval .= exists $EscapedTagToChar{$pair} ?
                  $EscapedTagToChar{$pair}
                  : substr $pair, 1, 1;
              } else {
                $realval .= $ch
              }
            }
          }
          $event{tags}->{$thistag} = $realval
    }
    $pos = $nextsp + 1;
  }

  $pos++ while substr($raw_line, $pos, 1) eq SPCHR;

  if ( substr($raw_line, $pos, 1) eq ':' ) {
    my $nextsp;
    ($nextsp = index $raw_line, SPCHR, $pos) > 0 and length(
        $event{prefix} = substr $raw_line, ($pos + 1), ($nextsp - $pos - 1)
    ) or return;
    $pos = $nextsp + 1;
    $pos++ while substr($raw_line, $pos, 1) eq SPCHR;
  }

  my $nextsp_maybe;
  if ( ($nextsp_maybe = index $raw_line, SPCHR, $pos) == -1 ) {
    # No more spaces; do we have anything..?
    my $cmd = substr $raw_line, $pos;
    $event{command} = uc( length $cmd ? $cmd : return );
    return \%event
  }

  $event{command} = uc( 
    substr($raw_line, $pos, ($nextsp_maybe - $pos) )
  );
  $pos = $nextsp_maybe + 1;

  $pos++ while substr($raw_line, $pos, 1) eq SPCHR;

  my $maxlen = length $raw_line;
  PARAM: while ( $pos < $maxlen ) {
    if ( substr($raw_line, $pos, 1) eq ':' ) {
      push @{ $event{params} }, substr $raw_line, ($pos + 1);
      last PARAM
    }
    if ( (my $nextsp = index $raw_line, SPCHR, $pos) == -1 ) {
      push @{ $event{params} }, substr $raw_line, $pos;
      last PARAM
    } else {
      push @{ $event{params} }, substr $raw_line, $pos, ($nextsp - $pos);
      $pos = $nextsp + 1;
      $pos++ while substr($raw_line, $pos, 1) eq SPCHR;
      next PARAM
    }
  }

  \%event
}


no bytes;


print
  qq[<mst> let's try this again -without- the part where we beat you to],
  qq[ death with a six foot plush toy of sexual harassment panda\n ]
unless caller; 1;


=pod

=head1 NAME

POE::Filter::IRCv3 - Fast IRCv3.2 parser for POE or stand-alone use

=head1 SYNOPSIS

  my $filter = POE::Filter::IRCv3->new(colonify => 1);

  # Raw lines parsed to hashes:
  my $array_of_refs  = $filter->get( 
    [ 
      ':prefix COMMAND foo :bar',
      '@foo=bar;baz :prefix COMMAND foo :bar',
    ]
  );

  # Hashes deparsed to raw lines:
  my $array_of_lines = $filter->put( 
    [
      {
        prefix  => 'prefix',
        command => 'COMMAND',
        params  => [
          'foo',
          'bar'
        ],
      },
      {
        prefix  => 'prefix',
        command => 'COMMAND',
        params  => [
          'foo',
          'bar'
        ],
        tags => {
          foo => 'bar',
          baz => undef,
        },
      },
    ] 
  );


  # Stacked with a line filter, suitable for Wheel usage, etc:
  my $ircd = POE::Filter::IRCv3->new(colonify => 1);
  my $line = POE::Filter::Line->new(
    InputRegexp   => '\015?\012',
    OutputLiteral => "\015\012",
  );
  my $stacked = POE::Filter::Stackable->new(
    Filters => [ $line, $ircd ],
  );

  # Functional parser interface:
  my $event = POE::Filter::IRCv3::parse_one_line(
    ':foo PRIVMSG #bar :baz quux'
  );

=head1 DESCRIPTION

A L<POE::Filter> for IRC traffic with support for IRCv3.2 message tags.

Does not rely on regular expressions for parsing.  Benchmarks show this
approach is generally faster on the most common IRC strings.

Like any proper L<POE::Filter>, there are no POE-specific bits involved here
-- the filter can be used stand-alone to parse lines of IRC traffic (also see
L<IRC::Toolkit::Parser>). 

In fact, you do not need L<POE> installed -- if L<POE::Filter> is not
available, it is left out of C<@ISA> and the filter will continue working
normally.

=head2 POE / Object interface

=head3 new

Construct a new Filter; if the B<colonify> option is true, 
the last parameter will always have a colon prepended.
(This setting can also be retrieved or changed on-the-fly by calling 
B<colonify> as a method, or changed for specific events by passing a 
B<colonify> option via events passed to L</put>.)

=head3 get_one_start, get_one, get_pending

Implement the interface described in L<POE::Filter>.

See L</get>.

=head3 get

  my $events = $filter->get( [ $line, $another, ... ] );
  for my $event (@$events) {
    my $cmd = $event->{command};
    ## See below for other keys available
  }

Takes an ARRAY of raw lines and returns an ARRAY of HASH-type references with 
the following keys:

=head4 command

The (uppercased) command or numeric.

=head4 params

An ARRAY containing the event parameters.

=head4 prefix

The sender prefix, if any.

=head4 tags

A HASH of key => value pairs matching IRCv3.2 "message tags" -- see 
L<http://ircv3.atheme.org>.

Note that a tag can be present, but have an undefined value.

=head3 put

  my $lines = $filter->put( [ $hash, $another_hash, ... ] );
  for my $line (@$lines) {
    ## Direct to socket, etc
  }

Takes an ARRAY of HASH-type references matching those described in L</get> 
(documented above) and returns an ARRAY of raw IRC-formatted lines.

=head4 colonify

In addition to the keys described in L</get>, the B<colonify> option can be 
specified for specific events. This controls whether or not the last 
parameter will be colon-prefixed even if it is a single word. (Yes, IRC is 
woefully inconsistent ...)

Specify as part of the event hash:

  $filter->put([ { %event, colonify => 1 } ]);

=head3 clone

Copy the filter object (with a cleared buffer).

=head3 debug

Turn on/off debug output, which will display every input/output line (and
possibly other data in the future).

This is enabled by default at construction time if the environment variable
C<POE_FILTER_IRC_DEBUG> is a true value.

=head2 Functional interface

=head3 parse_one_line

If the filter is being used as a stand-alone IRC parser and speed is of the
essence, you can skip method resolution & queue handling by calling the parse
function directly using the fully-qualified name:

  my $ev = POE::Filter::IRCv3::parse_one_line( $line );

The function takes a single line and returns a HASH whose structure is
described in the documentation for L</get>, above.

If the given line cannot be parsed, the function returns false (rather than
throwing an exception, as L</get> would).

There is currently no functional interface to message string composition
(L</put>).

=head1 AUTHOR

Jon Portnoy <avenj@cobaltirc.org>

Licensed under the same terms as Perl.

Original implementations were derived from L<POE::Filter::IRCD>, 
which is copyright Chris Williams and Jonathan Steinert. This codebase has
diverged significantly.

Major thanks to the C<#ircv3> crew on irc.atheme.org, especially C<Aerdan> and
C<grawity>, for various bits of inspiration.

=head1 SEE ALSO

L<IRC::Message::Object>

L<POE::Filter>

L<POE::Filter::IRCD>

L<POE::Filter::Line>

L<POE::Filter::Stackable>

L<IRC::Toolkit>

=cut
